### DISCUSSION TITLE
Formula with universal binary dependencies

### DISCUSSION THREAD

**Discussion by ldennington** [Homebrew/discussions#4479] [Writing Formulae/Casks]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.16-15-gfec24f5
ORIGIN: https://github.com/Homebrew/brew
HEAD: fec24f5c08b1792f96cefae13a69d0f2a9abdbab
Last commit: 7 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 72312c1f70f833c09ed2e741edd19cb2a74772dd
Core tap last commit: 24 hours ago
Core tap branch: update-git-credential-manager
Core tap JSON: 03 May 16:29 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.10 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 14.0.3 build 1403
Git: 2.40.0 => /usr/local/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-x86_64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3
```

### Output of `brew doctor`

```shell
N/A
```

### Description of issue

I am finishing up creation of a new formula for the [`git-credential-manager` project](https://github.com/git-ecosystem/git-credential-manager) (see related discussion [here](https://github.com/orgs/Homebrew/discussions/4460)). Everything is looking good _except_ that I'm seeing the following output when I run `brew audit --strict --online git-credential-manager`:

```
git-credential-manager:
  * Unexpected universal binaries were found.
    The offending files are:
      /usr/local/Cellar/git-credential-manager/2.1.0/libexec/libAvaloniaNative.dylib
      /usr/local/Cellar/git-credential-manager/2.1.0/libexec/libHarfBuzzSharp.dylib
      /usr/local/Cellar/git-credential-manager/2.1.0/libexec/libSkiaSharp.dylib
Error: 1 problem in 1 formula detected
```

The files in question are dependencies that are only offered as universal binaries. Is there a recommendation for projects that have these types of dependencies but would like to publish themselves as a formula?

**Comment by SMillerDev** [comment]
Formula shouldn't install prebuilt binaries to begin with: https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae

**Reply by ldennington** [reply]
I was under the impression that binary _dependencies_ are all right for formulae. I've found a couple examples that seem to support this:

1. [`jackett`](https://github.com/Homebrew/homebrew-core/blob/d1d271255f515c5cf39c4905be531d5bdf75cc62/Formula/jackett.rb#L19) - depends on [`Mono.Posix`](https://github.com/Jackett/Jackett/blob/master/src/Jackett.Server/Jackett.Server.csproj). `Mono.Posix` in turn depends on [`Mono.Unix`](https://www.nuget.org/packages/Mono.Posix#dependencies-body-tab), which [includes native binaries](https://nuget.info/packages/Mono.Unix/7.1.0-final.1.21458.1).

![image (5)](https://user-images.githubusercontent.com/11321782/236335288-63884758-b885-4e8a-a344-84a4139d8794.png)

These binaries are installed by the formula without being built from source. Since they are not universal, however, they don't run into the same issue we do with the `audit` command.

2. [`mono`](https://github.com/Homebrew/homebrew-core/blob/d1d271255f515c5cf39c4905be531d5bdf75cc62/Formula/mono.rb#L157) - clones the `msbuild` repo and runs a script to set it up, which [downloads a pre-built version of `msbuild`](https://github.com/mono/msbuild/blob/63458bd6cb3a98b5a062bb18bd51ffdea4aa3001/eng/cibuild_bootstrapped_msbuild.sh#LL80C23-L80C107). 

In the `dotnet` packaging system (NuGet), it’s typical to include native binaries in packages. So, in the case of GCM, the scenario is:

- We build GCM from source in our formula, using the [`dotnet` formula](https://github.com/Homebrew/homebrew-core/blob/8715c6ba6061a9988e3f91777221524a3007db3e/Formula/dotnet.rb) as a build dependency
- We reference the [`Avalonia.Desktop` NuGet package](https://github.com/git-ecosystem/git-credential-manager/blob/61cd8e37c2bd1416a981c6a048d15bf2e841400d/src/shared/Core/Core.csproj#L21) in one of our projects
- `Avalonia.Desktop` depends on [`Avalonia.Native`](https://www.nuget.org/packages/Avalonia.Desktop), which [contains native binaries](https://nuget.info/packages/Avalonia.Native/0.10.19).
- `Avalonia.Desktop` also depends on [`Avalonia.Skia`](https://www.nuget.org/packages/Avalonia.Skia/) which depends on [`SkiaSharp`](https://www.nuget.org/packages/SkiaSharp/) which depends on [`SkiaSharp.NativeAssets.macOS`](https://www.nuget.org/packages/SkiaSharp.NativeAssets.macOS/), which [contains native binaries](https://nuget.info/packages/SkiaSharp.NativeAssets.macOS/2.88.3).

Finally, it also looks like there are some existing formulae that [specifically install universal binaries](https://github.com/Homebrew/homebrew-core/blob/master/audit_exceptions/universal_binary_allowlist.json) - given the above, would it be possible for GCM to be added to this list?

**Reply by SMillerDev** [reply]
The rules are specified in the documentation. Something might slip through because our detection isn't infallible, but in principle we don't allow binaries.
